beacon_node/network/src/sync/network_context.rs

```diff
@@ -1449,6 +1733,175 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
+    /// If there are no peers to serve the column requests, we add them to a queue for retrying
+    /// the requests once more peers become available.
+    ///
+    /// Note: we do not use the by root syncing mechanism for backfill since there is only
+    /// one canonical chain to sync.
+    fn request_columns_on_successful_blocks(
+        &mut self,
+        id: BlocksByRangeRequestId,
+        blocks: &[Arc<SignedBeaconBlock<T::EthSpec>>],
+    ) -> Result<(), RpcResponseError> {
+        let batch_epoch = id.batch_id();
+        // Return early if no columns are required for this epoch
+        if !matches!(
+            self.batch_type(batch_epoch),
+            ByRangeRequestType::BlocksAndColumns
+        ) {
+            return Ok(());
+        }
+        // Return early if this is a backfill batch, backfill batches are handled by range requests instead of root
+        if matches!(
+            id.parent_request_id.requester,
+            RangeRequestId::BackfillSync { .. }
+        ) {
+            return Ok(());
+        }
+        debug!(count = blocks.len(), "Received blocks from byrange query");
+
+        let mut block_roots = Vec::new();
+
+        // We have blocks here, check if they need data columns and request them
+        for block in blocks.iter() {
+            // Request columns only if the blob_kzg_commitments is non-empty
+            if let Ok(commitments) = block.message().body().blob_kzg_commitments()
+                && !commitments.is_empty()
+            {
+                block_roots.push(block.canonical_root());
+            }
+        }
+
+        // No blobs for the entire epoch, let the coupling logic know not to expect anything
+        // and return early
+        if block_roots.is_empty() {
+            if let Some(req) = self
+                .components_by_range_requests
+                .get_mut(&id.parent_request_id)
+            {
+                if let Err(e) = req.no_columns_for_batch() {
+                    debug!(?e, "Created range request in inconsistent state");
+                    return Err(RpcResponseError::InternalError(e));
+                }
+                return Ok(());
+            } else {
+                return Err(RpcResponseError::InternalError(
+                    "Block request sent without creating a components_by_range entry".to_string(),
+                ));
+            }
+        }
+        // Generate the data column by root requests
+        let mut peer_to_columns: HashMap<PeerId, Vec<ColumnIndex>> = HashMap::new();
+        let mut no_peers_for_column: Vec<ColumnIndex> = Vec::new();
+        for column in self.chain.sampling_columns_for_epoch(batch_epoch).iter() {
+            let data_column = DataColumnSubnetId::new(*column);
+            if let Some(custody_peer) = self
+                .network_globals()
+                .peers
+                .read()
+                .good_custody_subnet_peer_range_sync(data_column, batch_epoch)
+                .next()
```

Comment:
This always picks the first matching peer, would it be possible for the same peer to keep getting selected for the same column across different batches?

Reply:
Its a hashmap so every iteration in the peerdb would return a different order with new peers getting added and old peers getting removed I think.